### ISSUE TITLE
Implement spinlock on top of atomic api and remove up_testset

### ISSUE THREAD

**Issue by xiaoxiang781216** [apache/incubator-nuttx#6301]
See the discussion here: https://github.com/apache/incubator-nuttx/pull/6283

**Comment by masayuki2009** [comment]
@xiaoxiang781216 
For example, cxd56xx needs to use up_testset API because the SoC has a restriction to use ldrex and strex.

**Comment by xiaoxiang781216** [comment]
Yes, in your case not only up_testset, but also atomic_xxx need rewrite for cxd56xx. The issue description is still right, it's better to use atomic api directly, but implement atomic api has different approach:

1. boolchain's bultin
2. libs/libc/machine/arc_atomic. for single core
3. chip special one(e.g. cxd56xx)

**Comment by xiaoxiang781216** [comment]
@hujun260 please look at this problem too.
